lib/srv/desktop/rdp/rdpclient/src/rdpdr/path.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::ffi::{CString, NulError};
+
+/// WindowsPath is a String that we assume to be in the form
+/// of a traditional DOS path:
+///
+/// https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats
+///
+/// Because RDP device redirection is limited in the paths it uses, we can
+/// further assume that it is in one of the following forms:
+///
+/// r"\Program Files\Custom Utilities\StringFinder.exe": An absolute path from the root of the current drive.
+///
+/// r"2018\January.xlsx": A relative path to a file in a subdirectory of the current directory.
+#[derive(Debug, Clone)]
+pub struct WindowsPath {
+    path: String,
+}
+
+impl WindowsPath {
+    pub fn new(path: String) -> WindowsPath {
+        Self { path }
+    }
+}
+
+/// UnixPath is a String that we assume to be in the form of a
+/// Unix Path, qualified by the qualifications laid out in RFD 0067
+///
+/// https://github.com/gravitational/teleport/blob/master/rfd/0067-desktop-access-file-system-sharing.md
+#[derive(Debug, Clone)]
+pub struct UnixPath {
+    pub path: String,
+}
+
+impl UnixPath {
+    pub fn new(path: String) -> UnixPath {
+        Self { path }
+    }
+
+    /// This function will create a CString from a UnixPath.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the UnixPath contains
+    /// any characters that can't be handled by CString::new().
+    pub fn as_cstring(&self) -> Result<CString, NulError> {
```

Comment:
```suggestion
    pub fn to_cstring(&self) -> Result<CString, NulError> {
```

as this is a (fallible, even) conversion that "takes effort".

Reply:
22a25408beee4f525ec3501739db1f8f224a5e75